subprojects/execution/src/test/groovy/org/gradle/internal/execution/history/impl/FileSystemSnapshotSerializerTest.groovy

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.history.impl
+
+import org.apache.commons.io.FilenameUtils
+import org.gradle.api.internal.cache.StringInterner
+import org.gradle.internal.file.FileMetadata.AccessType
+import org.gradle.internal.serialize.AbstractEncoder
+import org.gradle.internal.serialize.SerializerSpec
+import org.gradle.internal.snapshot.CompleteDirectorySnapshot
+import org.gradle.internal.snapshot.CompleteFileSystemLocationSnapshot
+import org.gradle.internal.snapshot.CompositeFileSystemSnapshot
+import org.gradle.internal.snapshot.RegularFileSnapshot
+
+import static java.lang.Math.abs
+import static org.gradle.internal.file.FileMetadata.AccessType.DIRECT
+import static org.gradle.internal.file.FileMetadata.AccessType.VIA_SYMLINK
+import static org.gradle.internal.file.impl.DefaultFileMetadata.file
+import static org.gradle.internal.hash.HashCode.fromInt
+import static org.gradle.internal.snapshot.FileSystemSnapshot.EMPTY
+
+class FileSystemSnapshotSerializerTest extends SerializerSpec {
+    def stringInterner = new StringInterner()
+    def serializer = new FileSystemSnapshotSerializer(stringInterner)
+    def pseudoRandom = new Random(1234L)
+
+    @Override
+    Class<? extends AbstractEncoder> getEncoder() {
+        return super.getEncoder()
+    }
+
+    def "reads and writes empty snapshots"() {
+        when:
+        def out = serialize(EMPTY, serializer)
+
+        then:
+        out == EMPTY
+    }
+
+    def "reads and writes file snapshots"() {
```

Comment:
There should be a test for writing missing snapshots as well.

Reply:
Added.